### PR TITLE
nix: factor out common.nix

### PR DESCRIPTION
**Pull request by lovesegfault**
###### Motivation for this change

This refactors the package a bit to make it easier to maintain. The main
advantage is it completely separates the package building logic from the
creation of new instances of Nix.

I also renamed `is24` and `is25` to `atLeast24` and `atLeast25` internally,
since that better represents what they are.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lovesegfault** [comment]
Result of `nixpkgs-review pr 156663` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>nix-doc</li>
    <li>nix-plugins</li>
    <li>vulnix</li>
  </ul>
</details>
<details>
  <summary>64 packages built:</summary>
  <ul>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cached-nix-shell</li>
    <li>cachix</li>
    <li>comma</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>hci</li>
    <li>hercules-ci-agent</li>
    <li>home-manager</li>
    <li>hubstaff</li>
    <li>hydra-unstable</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix</li>
    <li>nix-binary-cache</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-du</li>
    <li>nix-eval-jobs</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-unstable</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nix_2_3</li>
    <li>nix_2_4</li>
    <li>nix_2_5</li>
    <li>nixos-generators</li>
    <li>nixos-option</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>nvfetcher</li>
    <li>prefetch-yarn-deps</li>
    <li>python310Packages.nix-kernel</li>
    <li>python310Packages.nixpkgs</li>
    <li>python310Packages.pythonix</li>
    <li>python39Packages.nix-kernel</li>
    <li>python39Packages.nixpkgs</li>
    <li>python39Packages.pythonix</li>
    <li>rnix-lsp</li>
    <li>terranix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>yarn2nix</li>
  </ul>
</details>

**Comment by lovesegfault** [comment]
All failures happen on master too.

**Comment by lovesegfault** [comment]
Waiting on #156665 to rebase this.

**Comment by lovesegfault** [comment]
Result of `nixpkgs-review pr 156663` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>vulnix</li>
  </ul>
</details>
<details>
  <summary>66 packages built:</summary>
  <ul>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cached-nix-shell</li>
    <li>cachix</li>
    <li>comma</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>hci</li>
    <li>hercules-ci-agent</li>
    <li>home-manager</li>
    <li>hubstaff</li>
    <li>hydra-unstable</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix</li>
    <li>nix-binary-cache</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-eval-jobs</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-unstable</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nix_2_3</li>
    <li>nix_2_4</li>
    <li>nix_2_6</li>
    <li>nixos-generators</li>
    <li>nixos-option</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>nvfetcher</li>
    <li>prefetch-yarn-deps</li>
    <li>python310Packages.nix-kernel</li>
    <li>python310Packages.nixpkgs</li>
    <li>python310Packages.pythonix</li>
    <li>python39Packages.nix-kernel</li>
    <li>python39Packages.nixpkgs</li>
    <li>python39Packages.pythonix</li>
    <li>rnix-lsp</li>
    <li>terranix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>yarn2nix</li>
  </ul>
</details>

File: pkgs/tools/package-management/nix/patches/boehmgc-coroutine-sp-fallback.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+           /* FIXME: Need to scan the normal stack too, but how ? */
+           /* FIXME: Assume stack grows down */
++        } else {
++          if (pthread_getattr_np(p->id, &pattr)) {
++            ABORT("GC_push_all_stacks: pthread_getattr_np failed!");
++          }
++          if (pthread_attr_getstacksize(&pattr, &stack_limit)) {
++            ABORT("GC_push_all_stacks: pthread_attr_getstacksize failed!");
++          }
++          // When a thread goes into a coroutine, we lose its original sp until
```
Comment (mohe2015) [inline pkgs/tools/package-management/nix/patches/boehmgc-coroutine-sp-fallback.patch:25]: This patch is not up-to-date with https://github.com/NixOS/nix/blob/master/boehmgc-coroutine-sp-fallback.diff

Don't know if we should include that memory leak fix.

Comment (lovesegfault) [reply pkgs/tools/package-management/nix/patches/boehmgc-coroutine-sp-fallback.patch:28]: Updated the patch, let's see if everything still works.

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -661,7 +661,9 @@ mapAliases ({
   nilfs_utils = nilfs-utils; # added 2018-04-25
   nix-direnv-flakes = nix-direnv;
   nix-review = nixpkgs-review; # added 2019-12-22
-  nixFlakes = nixStable; # added 2021-05-21
+  nixFlakes = nix; # added 2021-05-21
+  nixStable = nix; # added-2022-01-24
+  nixUnstable = nix-unstable; # added 2022-01-24
```
Comment (roberth) [inline pkgs/top-level/aliases.nix:666]: This will cause a lot of warnings that are impossible to avoid without breaking compatibility with `nixos-22.05`. I've been guilty of this and it causes pain for out-of-tree expression maintainers as users complain about warnings.
Please move these to `all-packages.nix` with a comment suggesting to move them to `aliases` in 2023.

Did you rename these because we have a convention now? I might not have followed the news on this, but if we don't have a documented convention, we probably shouldn't rename these, as the next person could change it to their preference.

Comment (lovesegfault) [reply pkgs/top-level/aliases.nix:666]: I'm not sure I follow, why would make something an alias cause warnings? Moreover, why would the situation be better in 2023?

ISTR the convention is to use `kebab-case` for package names, but I don't remember where I got that from. I'm flexible on renaming `nixUnstable`, but IMHO `nixStable` should go; it's implied that `nix` is the stable release.

Comment (roberth) [reply pkgs/top-level/aliases.nix:666]: I was under the assumption that all definitions in this file caused a warning when used. They do not. Sorry for the noise.

`nixStable` could be useful if you want to upgrade `nix` in an overlay but need to make an exception for some packages.

> `kebab-case` for package names

The docs only specify `-`s for versions (`nix-2-5`) but it seems sensible to extrapolate this to version-like qualifiers.

**Comment by edolstra** [comment]
IMHO splitting it into multiple files makes it *harder* to maintain, not easier.

**Comment by roberth** [comment]
I think we should consider having a `nix-latest-release` because we haven't had a problem-free minor release yet but we're still calling new releases "stable". `nix` could lag behind that for about a week, assuming a sufficient number of people use `nix-latest-release`.

**Comment by edolstra** [comment]
I think we should consider removing `nix-unstable`. People who want the bleeding edge are better off using the `nix/master` flake, since it's always up to date.

**Comment by roberth** [comment]
> IMHO splitting it into multiple files makes it _harder_ to maintain, not easier.

I'm ok with `common.nix`, but having the `nix =` definition in `all-packages.nix` along with a bunch of duplication is annoying. It'd be nicer if those things were in an attrset together. `all-packages.nix` could just do `inherit nixVersions; nix = nixVersions.stable` "forever". Too radical?

**Comment by roberth** [comment]
> I think we should consider removing `nix-unstable`. People who want the bleeding edge are better off using the `nix/master` flake, since it's always up to date.

I'm not aware of any packages that use it as a dependency, except for `hydra`. That will have to be removed too.

**Comment by roberth** [comment]
Relevant for the naming convention change: https://github.com/NixOS/nixpkgs/issues/17625

I wonder why the documentation even claims `foo-1-2` is it. `[a-zA-Z]-[0-9]+(-[0-9]+)+ =` shows only 8 (eight!) results. I'd almost think we've been trolled.

**Comment by edolstra** [comment]
Yeah, removing Hydra (and using the upstream flake instead) would save a lot of work repackaging Hydra in Nixpkgs.

**Comment by SuperSandro2000** [comment]
> I think we should consider removing `nix-unstable`. People who want the bleeding edge are better off using the `nix/master` flake, since it's always up to date.

It still has url-literals in it which have been removed quite a while back.

File: pkgs/top-level/aliases.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/top-level/aliases.nix:666]: This rename is totally unnecessary and we haven't even agreed upon how to name it. It will also break things where alias are disabled like my local setup or nixpkgs-review. If the rename is kept in all-packages.nix until next stable release there is some time to adjust to it. Also please move that to another PR since it is a bigger topic and we could then move on (or not) with the extraction.

**Comment by edolstra** [comment]
I think the 6-week release schedule makes it a lot less important to have a `nix-unstable` in Nixpkgs.

Not sure what the relevance of URL literals is. They've not been removed from the Nix language so it's fine to use them in your flakes.

File: pkgs/top-level/aliases.nix (hunk 2 above)
Comment (roberth) [reply pkgs/top-level/aliases.nix:666]: Please use underscores and do it in a separate PR. 
Sorry for the 180 on this, but the data speaks for itself, at least for version numbers. See https://github.com/NixOS/nixpkgs/pull/156857. 
Qualifiers are hard to grep for, but the version to qualifier extrapolation argument still makes sense.

Comment (lovesegfault) [reply pkgs/top-level/aliases.nix:666]: Moving the `nixUnstable` -> `nix_unstable` rename to a subsequent PR.

**Comment by lovesegfault** [comment]
@roberth Alright, I liked your suggestions of `nixVersions` so see how it looks now.

TL;DR: 
* `callPackage tools/package-management/nix` returns an attrset of all nix versions (`nix_2_3`, ..., `nix_2_6`, `stable`, `unstable`)
* `all-packages.nix` puts that in a `nixVersions` attr at the top-level.
* We define `nix = nixVersions.stable` at the top-level
* `nixUnstable` is made an alias of `nixVersions.unstable`

Much less duplication all around

**Comment by lovesegfault** [comment]
sync'd with rob out-of-band, AOK, landing

**Comment by edolstra** [comment]
Sorry but I really don't see how scattering code across a default.nix and common.nix is an improvement. That just makes the code harder to read since you have to constantly switch between mutliple files. It also leads to duplication, e.g. the aws-sdk-cpp argument is an argument to both files.

It's also really weird to have the boehm-gc package depend on `nixVersions`. And having Hydra take `nixVersions` as an argument is poor style since it conveys that it depends on all versions of Nix. It's better to depend on just `nix-unstable`.

**Comment by lovesegfault** [comment]
> It's also really weird to have the boehm-gc package depend on `nixVersions`. 

That was copying what the code was doing previously since we had every Nix variant in boehmgc's passthru.tests. I'm not sure why we had that to begin with, but I chose not to alter it.

> And having Hydra take `nixVersions` as an argument is poor style since it conveys that it depends on all versions of Nix. It's better to depend on just `nix-unstable`.

Fixing in #157072

**Comment by SuperSandro2000** [comment]
This also breaks simple overlays when nixUnstable was used. Now you need to append it to nixVersions.

**Comment by lovesegfault** [comment]
> This also breaks simple overlays when nixUnstable was used. Now you need to append it to nixVersions.

The alternative was to keep having to update the list of maintained Nix versions in three places on every bump which is a PITA. 

We could make a `nix_unstable` top-level attr again that alised `nixVersions.unstable` which would make the case of wanting to override `nix_unstable` more ergonomic, but would do nothing for the other `nix_X_Y` variants. 

Another idea to make the use-case easier is to call `makeExtensible` on `nixVersions` which would let you write

```
_: prev: {
  nixVersions = prev.nixVersions.extend(_: super: {
    unstable = super.unstable.overrideAttrs (_: { });
  });
}
```

Which IMO is a bit better.

**Comment by roberth** [comment]
`nix_unstable` has become irrelevant anyway, as we're unlikely to need it between the now-frequent releases. When we do need it, it seems better to use the flake (or "legacy" import) instead, so we get the exact version we want.

That is, unless we want to call the latest release "unstable", which given the current track record isn't a bad idea.

**Comment by SuperSandro2000** [comment]
> Another idea to make the use-case easier is to call `makeExtensible` on `nixVersions` which would let you write

If that is not written into the documentation like this, no normal or advanced user will be able to come up with this in a reasonable amount of time.

> nix_unstable has become irrelevant anyway, as we're unlikely to need it between the now-frequent releases. 

If certain PRs still take months to merge it is not irrelevant which is certainly the case.

> When we do need it, it seems better to use the flake (or "legacy" import) instead, so we get the exact version we want.

If you are not using flakes you're uselessly duplicating a lot of core libraries and deployment from multiple to multiple machines becomes unnecessarily hard and you're also wasting storage. Switching to flakes on the other hand is a very time consuming and frustrating process and took me the better part of two weeks which I can't really recommend to anyone.

**Comment by zowoq** [comment]
https://hydra.nixos.org/eval/1740837#tabs-removed

Looks like the various `nix` versions (`nixUnstable`, `nix_2_*`) are no longer evaluated (as expected) but `nixVersions.*` doesn't seem to be evaluated either? So they are only being built if they are the closure of another package that is built on hydra?

**Comment by lovesegfault** [comment]
> https://hydra.nixos.org/eval/1740837#tabs-removed
> 
> Looks like the various `nix` versions (`nixUnstable`, `nix_2_*`) are no longer evaluated (as expected) but `nixVersions.*` doesn't seem to be evaluated either? So they are only being built if they are the closure of another package that is built on hydra?

Good catch, handled here: https://github.com/NixOS/nixpkgs/pull/157210
